src/data/array/defs.lean

```diff
@@ -0,0 +1,54 @@
+/-
+Copyright (c) 2018 Keeley Hoek. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Keeley Hoek
+-/
+
+universes u v z
+
+namespace array
+
+-- TODO There's no need for these to be meta; if someone needs them in proof land
+-- they can add proofs that the inductions are well-founded.
+
```

Comment:
I think this is a repetition of the `traversable` instance for `array`

Reply:
I couldn't find such an instance. I don't really know what `traversable` is for or how to use it. Could you give me a hint what we should do here?

Reply:
(No prejudgement but) I'd also just note that we'd like this to be as fast as possible.

Reply:
If my instance is not fast enough, you can make it faster. Others might benefit from that. The instance itself is in `data.array.lemmas`

Reply:
It is also more general: it requires an applicative, not a monad.

Reply:
If you decide to make it faster, I recommend you prove that whatever you write gives the same result as `@equiv.traversable (flip vector n)`. Then, you get the laws for free.

Reply:
Ugh... I'm really not sure what to do here. I just want
```
def list_map_copy {α : Type u} {β : Type v} {n : ℕ} (x : array n β) (fn : α → β) (l : list α) := ...
```
to be available, and to run quickly. I don't know what `traversable` is for, or how to use a `traversable` instance to write this function. :-(

I made this PR separately in the hope of chopping up our mega-PR on `rewrite_search` into smaller pieces, but the end result has just been getting stuck.